cpp/ql/src/semmle/code/cpp/AutogeneratedFile.qll

```diff
@@ -3,11 +3,45 @@ import semmle.code.cpp.File
 import semmle.code.cpp.Preprocessor
 
 /**
- * Holds if `c` is a comment which is usually seen in autogenerated files.
- * For example, comments containing 'autogenerated' or 'generated by'.
+ * Holds if comment `c` indicates that it might be in an auto-generated file, for
+ * example because it contains the text "auto-generated by".
  */
-predicate isAutogeneratedComment(Comment c) {
-  c.getContents().regexpMatch("(?si).*(?:auto[ -]?generated|generated (?:by|file)|changes made in this file will be lost).*")
+private bindingset[comment] predicate autogeneratedComment(string comment) {
+  // ?s = include newlines in anything (`.`)
+  // ?i = ignore case
+  exists(string cond |
+    cond =
+      // generated by (not mid-sentence)
+      "(^ generated by[^a-z])|" +
+      "(! generated by[^a-z])|" +
+
+      // generated file
+      "(generated file)|" +
+
+      // file [is/was/has been] generated
+      "(file( is| was| has been)? generated)|" +
+
+      // changes made in this file will be lost
+      "(changes made in this file will be lost)|" +
+  
+      // do not edit/modify
+      "(^ do(n't|nt| not) (hand-?)?(edit|modify))|" +
+      "(! do(n't|nt| not) (hand-?)?(edit|modify))" and
+  
+    comment.regexpMatch("(?si).*(" +
+      // replace `generated` with a regexp that also catches things like
+      // `auto-generated`.
+      cond.replaceAll("generated", "(auto[\\w-]*[\\s/\\*\\r\\n]*)?generated")
+
+      // replace `!` with a regexp for end-of-sentence / separator characters.
+      .replaceAll("!", "[\\.\\?\\!\\-\\;\\,]")
+
+      // replace ` ` with a regexp for one or more whitespace characters
+      // (including newlines and `/*`).
+      .replaceAll(" ", "[\\s/\\*\\r\\n]+") +
+      ").*"
```

Comment:
As a long-time Perl programmer, I am in awe of this use of a regex to transform another regex!

Reply:
It's actually only a string substitution on a regexp string, not a regexp on a regexp.  Things were getting surprisingly unreadable until I did it this way!